esp-hal/src/system.rs

```diff
@@ -187,14 +192,19 @@ impl Drop for PeripheralGuard {
 pub(crate) struct GenericPeripheralGuard<const P: u8> {}
 
 impl<const P: u8> GenericPeripheralGuard<P> {
-    pub(crate) fn new() -> Self {
+    pub(crate) fn new_with(init: fn()) -> Self {
         let peripheral = unwrap!(Peripheral::try_from(P));
         if !KEEP_ENABLED.contains(&peripheral) && PeripheralClockControl::enable(peripheral) {
             PeripheralClockControl::reset(peripheral);
+            init();
```

Comment:
This is what I meant when I said splitting the channels upfront is prone to init race conditions.

The init sequence here is:
1. enable if needed
2. reset and init if it wasn't enabled

The deinit sequence is:
1. disable

What happens when (on two different cores) the deinit occurs in between step 1 and 2?

We end up with a disabled channel in a driver.

Reply:
> What happens when (on two different cores) the deinit occurs in between step 1 and 2?

Deinit decrements the refcount by 1 and does nothing else.

Reply:
That being said, two cores enabling DMA at the same time IS somewhat racey, as the slower core may return and assume the DMA is initialized while the other core is running `init()`.

Reply:
I've convinced myself this is correct again. Since there is no further tear down happening after disabling, deinit won't race against init.

However, what happens when two inits are happening at the same time?

One core will "win" and do the bump from 0 -> 1.

If the winning core is interrupted in between step 1 and 2. The loosing core will skip step two and start using the channel assuming that the winning core has done step 2.
The loosing core is now using a channel that hasn't been pulled out of reset, or using a channel that will eventually be reset by the winning core when it's interruption is done.

"core" is interchangeable with "embassy task in interrupt executor" I think.